test/Database/LSMTree/Model/Session.hs

```
@@ -246,59 +250,60 @@ runModelMWithInjectedErrors ::
-        showString "ErrSnapshotExists"
-      ErrSnapshotDoesNotExist ->
-        showString "ErrSnapshotDoesNotExist"
-      ErrSnapshotWrongType ->
-        showString "ErrSnapshotWrongType"
-      ErrBlobRefInvalidated ->
-        showString "ErrBlobRefInvalidated"
-      ErrCursorClosed ->
-        showString "ErrCursorClosed"
-      ErrSnapshotCorrupted s ->
-        showParen (d > appPrec) $
-        showString "ErrSnapshotCorrupted " .
-        showParen True (showString s)
-      ErrDiskFault s ->
-        showParen (d > appPrec) $
-        showString "ErrDiskFault " .
-        showParen True (showString s)
-      ErrOther s ->
-        showParen (d > appPrec) $
-        showString "ErrOther " .
-        showParen True (showString s)
+  | ErrDiskFault !String
+  | ErrCommitActionRegistry (NonEmpty Err)
+  | ErrAbortActionRegistry (NonEmpty Err)
+  | ErrOther !String
+  deriving stock (Eq, Show)
+
+isSnapshotCorrupted :: Err -> Bool
+isSnapshotCorrupted (ErrDiskFault _)             = True
+isSnapshotCorrupted (ErrSnapshotCorrupted _)     = True
+isSnapshotCorrupted (ErrCommitActionRegistry es) = firstNotOtherSatisfies isSnapshotCorrupted es
+isSnapshotCorrupted (ErrAbortActionRegistry es)  = firstNotOtherSatisfies isSnapshotCorrupted es
+isSnapshotCorrupted _                            = False
+
+isDiskFault :: Err -> Bool
+isDiskFault (ErrDiskFault _)             = True
+isDiskFault (ErrCommitActionRegistry es) = firstNotOtherSatisfies isDiskFault es
+isDiskFault (ErrAbortActionRegistry es)  = firstNotOtherSatisfies isDiskFault es
+isDiskFault _                            = False
+
+isOther :: Err -> Bool
+isOther (ErrOther _)                 = True
+isOther (ErrCommitActionRegistry es) = all isOther es
+isOther (ErrAbortActionRegistry es)  = all isOther es
+isOther _                            = False
+
+firstNotOtherSatisfies :: (Err -> Bool) -> NonEmpty Err -> Bool
+firstNotOtherSatisfies p =
+  maybe False (p . fst) . L.uncons . NE.dropWhile isOther
```

I wonder if we shouldn't ignore `ErrOther`. The way I see it, `ErrOther` is always an unexpected exception, so maybe it should always lead to a test failure, which would be contingent on not "swallowing" the `ErrOther` in `isDiskFault` and `isSnapshotCorrupted`

---

Similarly, I think `isOther` should hold when `isOther` holds for *any* of exceptions recursively

---

I could make that change but I need the current behavior, so it'd be kinda pointless?

---

If I change it to:

```haskell
isSnapshotCorrupted :: Err -> Bool
isSnapshotCorrupted (ErrDiskFault _)             = True
isSnapshotCorrupted (ErrSnapshotCorrupted _)     = True
isSnapshotCorrupted e@(ErrCommitActionRegistry es) = not (isOther e) && (isSnapshotCorrupted $ fst $ NE.uncons es)
isSnapshotCorrupted e@(ErrAbortActionRegistry es)  = not (isOther e) && (isSnapshotCorrupted $ fst $ NE.uncons es)
isSnapshotCorrupted _                            = False

isDiskFault :: Err -> Bool
isDiskFault (ErrDiskFault _)             = True
isDiskFault e@(ErrCommitActionRegistry es) = not (isOther e) && (isDiskFault $ fst $ NE.uncons es)
isDiskFault e@(ErrAbortActionRegistry es)  = not (isOther e) && (isDiskFault $ fst $ NE.uncons es)
isDiskFault _                            = False

isOther :: Err -> Bool
isOther (ErrOther _)                 = True
isOther (ErrCommitActionRegistry es) = any isOther es
isOther (ErrAbortActionRegistry es)  = any isOther es
isOther _                            = False
```

then the tests pass for me. Do you need the previous behaviour for something else?

---

I've suggested in an earlier comment to bring it back

---

Do you want me to revert 17e8de75f5e2cf2c55a95b639b7f2efaaae9219d?

---

> If I change it to:
> 
> ```haskell
> isSnapshotCorrupted :: Err -> Bool
> isSnapshotCorrupted (ErrDiskFault _)             = True
> isSnapshotCorrupted (ErrSnapshotCorrupted _)     = True
> isSnapshotCorrupted e@(ErrCommitActionRegistry es) = not (isOther e) && (isSnapshotCorrupted $ fst $ NE.uncons es)
> isSnapshotCorrupted e@(ErrAbortActionRegistry es)  = not (isOther e) && (isSnapshotCorrupted $ fst $ NE.uncons es)
> isSnapshotCorrupted _                            = False
> 
> isDiskFault :: Err -> Bool
> isDiskFault (ErrDiskFault _)             = True
> isDiskFault e@(ErrCommitActionRegistry es) = not (isOther e) && (isDiskFault $ fst $ NE.uncons es)
> isDiskFault e@(ErrAbortActionRegistry es)  = not (isOther e) && (isDiskFault $ fst $ NE.uncons es)
> isDiskFault _                            = False
> 
> isOther :: Err -> Bool
> isOther (ErrOther _)                 = True
> isOther (ErrCommitActionRegistry es) = any isOther es
> isOther (ErrAbortActionRegistry es)  = any isOther es
> isOther _                            = False
> ```
> 
> then the tests pass for me. Do you need the previous behaviour for something else?

This has rather different semantics. The point is to find the first cause that isn't an `ErrOther` and classify that. This handles the scenario where the causes contain both corruption and disk faults.

The semantics of `isOther` are also incorrect here. `ErrOther` is the bottom element of a lattice, so any non-other error takes precedence, which explains the need for `all`. Corruption and disk fault on the other hand take precedence (in order) over any other error.

---

Probably, yes

---

> Probably, yes

W.r.t. reverting or w.r.t. agreeing that these predicates should indeed use the causes of action registry errors differently?

---

Reverting